feat(Slideovers): Allow multiple instances of programmatic slideover

genu opened the pull request:
<!---
☝️ PR title should follow conventional commits (https://conventionalcommits.org)
-->

### 🔗 Linked issue

<!-- If it resolves an open issue, please link the issue here. For example "Resolves #123" -->

### ❓ Type of change

<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->

- [ ] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description
This PR opens the possibility to open multiple instance of the `<Slideover/>` component using the `useSlideover()` composable.

This capability was discussed in the past, so this is an attempt at such a feature.

### 📝 Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have linked an issue or discussion.
- [x] I have updated the documentation accordingly.

genu committed:
feat(Slideover): Allow opening multiple independent instances

genu committed:
docs(Slideover): Update example to demonstrate multiple instances

genu committed:
chose(Slideover): Cleanup

genu commented:
Tagging @noook  for feedback

genu committed:
chore: Cleanup playground

genu committed:
chore: restore playground app.vue

genu committed:
refactor(Slideovers): Update Slideover component and related files

genu commented on src/runtime/composables/useSlideover.ts:
```diff
@@ -1,63 +1,84 @@
+        component,
+        props: props ?? {},
+        close: () => close(id)        
+      } 
```
I'm assuming `patch` should also be added to the instance to be able to do?:

```ts
instance.patch({...})
```

But i'm not sure how to handle the generic type

genu commented on src/runtime/composables/useSlideover.ts:
```diff
@@ -1,63 +1,84 @@
     reset,
-    patch,
-    isOpen
+    patch
```
These can still be useful to be able to modify all instances using the composable. as all of them (with the exception of `open` require an instance `id`)

genu commented on src/runtime/composables/useSlideover.ts:
```diff
@@ -1,63 +1,84 @@
+        isOpen: true,
+        component,
+        props: props ?? {},
+        close: () => close(id)        
```
Would it make sense to add `reset` on the instance?

It feels like `reset` is primarily used internally, and it would be hardly helpful in user land

noook replied:
I don't think it should

genu commented on src/runtime/types/slideover.d.ts:
```diff
@@ -10,7 +10,10 @@ export interface Slideover {
   component: Component | string
   props: Slideover
-}
+  close: () => void,
```
if `patch` is added here, how would I type it to be able to get inference for the proper component context? 🤔

noook replied:
I think something like this should do the trick

```ts
export interface SlideoverInstance<T> {
  id: number,
  isOpen: boolean
  component: Component | string
  props: Slideover & T
  patch: (props: Partial<Slideover & T>) => void
  close: () => void
}

function createInstance<T extends Component> (component: T, props?: Slideover & ComponentProps<T>): SlideoverInstance<ComponentProps<T>>`

const appSlideover = slideover.createInstance(AppSlideover, {
  title: 'Title'
})

appSlideover.patch({
  title: 'Updated title',
})
```

Update:
You'd need to type `open` the same way

genu replied:
In the current implementation we don't actually export the `createInstance` method. Its used internally when calling `instance.open(...)` and an instance is returned from `open()`.

noook replied:
Yes that's why you should type at least `open`, typing `createInstance` will just make it "match"

genu committed:
refactor(Slideover): Rename state to instance

genu committed:
refactor(Slideover): update

noook commented on src/runtime/composables/useSlideover.ts:
```diff
@@ -1,63 +1,84 @@
 
+  function createInstance<T extends Component> (component: T, props?: Slideover & ComponentProps<T>): SlideoverInstance {
+    // Random short id
+    const id = Math.floor(Math.random() * 1000000)
```
I think it would be safer to use `useId()`

genu replied:
I tried to use it, but it throws an error:
`[nuxt] useId must be called within a component setup function.` 🤔

noook commented on src/runtime/composables/useSlideover.ts:
```diff
@@ -1,63 +1,84 @@
 
-  async function close () {
-    if (!slideoverState) return
+  async function close (id: number) {
```
It doesn't need to be async

noook commented on src/runtime/composables/useSlideover.ts:
```diff
@@ -1,63 +1,84 @@
-      component: 'div',
-      props: {}
-    }
+  function reset (id: number) {     
```
`reset` should now be `remove`

noook commented on src/runtime/composables/useSlideover.ts:
```diff
@@ -1,63 +1,84 @@
+        id,
+        isOpen: true,
+        component,
+        props: props ?? {},
```
This will allow the users not to manually have to close the instance all the time
```suggestion
        props: {
          ...(props ?? {}),
          onClose() {
            props?.onClose?.()
            close(id)
          },
        },
```

genu replied:
Would we need to also add the `onClose` to the  to the `props` interface in `slideover.d.ts`?

noook replied:
Yes you can, in this context the typings can't detect the `T` component is a `USlideover`.

When I did this feature for modals I tried constraining `T` to be a subset of `<UModal>` but the types were not working correctly so I just restrained it to extend a regular `Component`. For the moment you can just port the events the `Slideover` emits to the `Slideover` type !

noook commented on docs/components/content/examples/SlideoverExampleComponent.vue:
```diff
@@ -1,19 +1,40 @@
+  const instance = slideover.open(SlideoverExampleComponentB, {
+    count: anotherCount.value,
+    side: 'left',
+    onClose: () => {
+      instance.close()
+    }
```
I don't think it's a good idea to oblige the user to write this same code all the time, I'll give you a snippet below

noook commented on src/runtime/components/overlays/Slideovers.client.vue:
```diff
@@ -1,18 +1,19 @@
-    @after-leave="reset"
+    :is="instance.component"
+    v-for="(instance, index) in slideoverInstances"
+    :key="index"
```
I'd suggest using the instance ID. Also, you update the list using `filter` and reinserting the new element at the end, which will change the item's indexes.

```suggestion
    :key="instance.id"
```

noook commented on src/runtime/composables/useSlideover.ts:
```diff
@@ -1,63 +1,84 @@
+      {
+        ...slideoverInstance,        
+        isOpen: false      
+      }
```
You already update `isOpen` a few lines above
```suggestion
      slideoverInstance
```

genu committed:
refactor(Slideover): Use instance id

genu committed:
refactor(Slideover): rename `reset` to `remove`

genu committed:
refactor(Slideover): Remove unnecessary code in useSlideover.ts

genu committed:
refactor(Slideover): Simplify useSlideover.ts code

genu committed:
refactor(Slideover): Don't make close async

genu committed:
refactor(Slideover): Cleanup

genu committed:
refactor(Slideover): Improve useSlideover.ts code readability and maintainability

genu committed:
Merge branch 'dev' into feature/multi-slideover

genu commented:
@noook Thanks for the comments, I made most changes.

I'm thinking, would it make better sense to create instances without `open`?

This would allow us do declare instances like this:

```ts
const slideover = useSlideover()
const slideoverA = slideover.create(SlideoverExampleComponent)

const openSlideover = () => {
  const slideoverA = slideover.open()
}

const doSomethingElse = () => {
   slideOverA.patch({..})
}

const doAnotherThing = () => {
   slideOverA.close()
}
```

noook commented:
There is an issue: Slideovers should be dismissable when clicking outside **unless** you explicitly prevent them from closing with the appropriate prop.

Your example components do not specify it so it should keep the original behaviour:
https://ui-9x1er5zyw-nuxt-js.vercel.app/dev/components/slideover#control-programmatically

genu committed:
refactor(Slideover): Add close event listener for slideover instances

genu committed:
Merge branch 'dev' into feature/multi-slideover

genu committed:
refactor(Slideovers): Remove unused code

The code changes in `Slideovers.client.vue` remove the unused `<pre>` element that displayed the `slideoverInstances` props. This cleanup improves code readability and removes unnecessary code.

genu commented:
> There is an issue: Slideovers should be dismissable when clicking outside **unless** you explicitly prevent them from closing with the appropriate prop.
> 
> Your example components do not specify it so it should keep the original behaviour: https://ui-9x1er5zyw-nuxt-js.vercel.app/dev/components/slideover#control-programmatically

I fixed it by listening to model changes on the modal and closing it when the model is changed. An issue I run into is that when clicking on the outside, it always closes the first one that was open, rather than the last one (i.e. the one on top). 

This issue would probably happen if we implement something similar in `UModal`. I'm digging through how the outside click is handled in the component and see if there is anything to be done there.

@noook Open to suggestions.

genu committed:
Merge branch 'dev' into feature/multi-slideover

benjamincanac commented:
Would it be possible to apply the same changes to the `useModal` composable in the same PR? @noook What do you think?

genu committed:
Merge branch 'dev' into feature/multi-slideover

# Conflicts:
#	docs/components/content/examples/SlideoverExampleComponent.vue

genu committed:
fix: Handle the preventClose event

genu commented:
@benjamincanac  I updated this PR to move it along.

A caveat that I ran into, (maybe its not a major issue?) is the `preventClose` attribute (which defaults to `false`) affects all of the slideover instances.

What this means in practice is that when `preventClose` is set to false, all of the slideovers on the page would respond to the close event, becaues there is no way to know "which" slideover's backdrop actually triggered the close event.

I don't think there is a way to get around this limitation at the moment without doing some gymnastics around the `headlessui` components.

We can document this limitation and roll with it, or wait until the next major release and deal with it then when we move to `radix-vue` 

Let me know your thoughts.

If we decide to roll with it, I can make the same changes to the `useModal` composable.

genu committed:
chore: restore app.vue

genu committed:
chore: fix formatting in app.vue

genu committed:
Merge branch 'dev' into feature/multi-slideover

benjamincanac commented:
@genu Sorry for the delay!! Should we really release this with the Headless UI limitation though?

I would rather implement this only in the `v3`.

genu commented:
I agree, not a big fan releasing a half baked solution.

I'm fine with revisiting this feature with the release of v3.

As a side note, I have been playing with `vue-final-modal` as a drop replacement for both `useModal` and `useSlideover` and it's been working much better than the built in compostables.

Perhaps for v3, it can be used as an inspiration.

benjamincanac commented:
Should we close this then? If you want to work on this, you can make a pull request against the `v3` branch already https://github.com/nuxt/ui/tree/v3 😊
